### PR TITLE
New action "edit-contacts" supports editing contacts across the hierarchy

### PR DESCRIPTION
**Pull request by rosteve**
At least once a week, the support team normally gets a couple of requests to bulk-edit contacts to add one or many attributes. We mostly use custom python/NodeJS/ and other bash scripts that we've developed to automate these tasks.
 
Most of the tasks we normally get on a weekly or sometimes daily basis include:
1. Updating supervisors. 
2. Tagging CHPs on a new workflow so that we can pick the attribute using the form context so that forms may be shown based on the CHW tag added to their documents.
3.  Bulk deleting CHWs or Forms that have been wrongly uploaded. 
4.  Adding some partner-specific attributes to specific docs in the medic CouchDB database such as external_id that may not have been added before during the user creation process.

Roughly these are the logical steps used to edit contacts: 
1. Validate the user doesn't update any specially reserved keyword.
2. Read the CSV to get the IDs to use to fetch the documents from CouchDB.
3. Check if the user specified the valid columns that should be added to the docs.
4. We edit/add the attributes specified to the doc and write all changes to disk for the user to check/confirm before uploading.
5. Run the upload-docs command to commit changes to the server.

@IreneAchola and I created this script and we've borrowed most of the logic from `csv-to-docs` and `move-contacts` which has already done most things we needed for this cc: @kennsippell .  We've tested the script manually on almost all scenarios we may need the script for and made changes to it to be as generic to most cases as possible.

How to configure this and an example for this can be found in the README.

The hope is that this tool can serve to replace those scripts that the support and tech lead team develop to bulk-edit contacts. cc: @smbuthia 
Solving issue #297

### CODE REVIEW

**Comment by garethbowen** [comment]
Awesome work, thanks for this!

@dianabarsan Do you have time to review this?

File: src/fn/edit-contacts.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  fs.mkdir(docDirectoryPath);
+  const saveJsonDoc = doc => fs.write(`${docDirectoryPath}/${doc._id}.doc.json`, toDocs.toSafeJson(doc) + '\n');
+
+  const csvDir = `${environment.pathToProject}/csv`;
+  if(!fs.exists(csvDir)) {
+    warn(`No csv directory found at ${csvDir}.`);
+    return Promise.resolve();
+  }
+
+  return fs.recurseFiles(csvDir)
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: It looks like you are iterating over every .csv file in the "csv" folder and throwing an error when there is at least one file that doesn't start with "contact.".
Is your intent is to support having multiple CSV files be processed at once: `contact._1.csv`, `contact.s.csv`, `contacts._223.csv` ? 
The code would be much cleaner if you just read a `contact.csv` file and process it - just like your example from the documentation - and run the command multiple times over different files, if needed.

File: src/fn/edit-contacts.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+  .then(() => model.exclusions.forEach(toDocs.removeExcludedField))
+  .then(() => Promise.all(Object.values(model.docs).map(saveJsonDoc)));
+};
+
+const model = {
+  csvFiles: {},
+  docs: {},
+  exclusions: [],
+  users: []
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: As far as I can tell, `users` property is not used.

File: src/fn/edit-contacts.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+function getIDs(csv) {
+    const { rows, cols } = fs.readCsv(csv);
+    var index = cols.indexOf('uuid');
+    if (index === -1){
+     throw Error('missing "uuid" column.');
+   }
+   return rows.map((item) => item[index]);
+}
+
+function processContacts(contactType, csv, ids, contactDocs, args){
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: I don't think you need to `contactType` parameter here. It's hardcoded to `contact` and then never used in the next function that it's passed to.

File: src/fn/edit-contacts.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  const { rows, cols } = fs.readCsv(csv);
+  var colNames = args.colNames;
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: You are assigning `colNames` here only to reassign the value later to either `cols` or a subset of `cols`.
I found it pretty hard to follow when reading this code and kept mixing `colNames` and `cols`.
Consider using more descriptive names like "csvColumns" and "columnsToEdit" or something similar and not reassigning variables to represent different things.

File: src/fn/edit-contacts.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  var index = cols.indexOf('uuid');
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: Consider using a meaningful name here like `uuidIndex` or similar.

File: src/fn/edit-contacts.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  var toIncludeIndex = [];
+  if (colNames.length === 0) {
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: Consider using shorthand `if (!colNames.length)`

File: src/fn/edit-contacts.js [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    var fileColumn;
+    for(fileColumn of cols) {
+      if (fileColumn !== null) {
+        splitFileColumns.push(fileColumn.split(':')[0]);
+      }      
+    }
+    const columnsValid = colNames.every(function(column) { return splitFileColumns.includes(column);});
+    if (!columnsValid){
+      throw Error('The column name(s) specified do not exist.');
+    }
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: As far as I can tell, this part only evaluates whether every `args.colNames` exists as a column in the csv.
Consider pulling this out as a separate function that returns a Boolean that represents whether the `args.colNames` are valid.

File: src/fn/edit-contacts.js [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    warn(' No columns specified, the script will add all the columns in the CSV!');
+    colNames = cols;
+  } else {
+    var splitFileColumns = [];
+    var fileColumn;
+    for(fileColumn of cols) {
+      if (fileColumn !== null) {
+        splitFileColumns.push(fileColumn.split(':')[0]);
+      }      
+    }
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: this can be rewritten as
```
const splitFileColumns = cols.map(col => col && col.split(':')[0]);
```

File: src/fn/edit-contacts.js [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    const columnsValid = colNames.every(function(column) { return splitFileColumns.includes(column);});
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: Use arrow functions for one liners:
```
colNames.every(column => splitFileColumns.includes(column));
```

File: src/fn/edit-contacts.js [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    if (!columnsValid){
+      throw Error('The column name(s) specified do not exist.');
+    }
+
+    colNames = cols.filter(e => colNames.includes(e.split(':')[0]));
+    colNames.forEach(function(e){
+      toIncludeIndex.push(cols.indexOf(e));
+    });
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: Consider rewriting this as:
```
const toIncludeIndex = colNames.map(col => cols.indexOf(col));
```

File: src/fn/edit-contacts.js [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+  }
+  if (colNames.includes('uuid')) {
+    colNames.splice(colNames.indexOf('uuid'),1);
+  }
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: You  need to exclude the `uuid` from `toIncludeIndex` as well, if I use `uuid` as a column to be edited, all values will be offset.
Consider adding some protected fields here, additional to the ones protected in `csv-to-docs`, like `parent` or `contact_type` or `contact`.

File: src/fn/edit-contacts.js [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+  return rows
+  .map(r => processCsv(contactType, colNames, r, ids, index, toIncludeIndex, contactDocs));
+
+}
+
+function processCsv(docType, columnNames, rowItem, ids, index, toIncludeIndex, contactDocs) {
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: `docType` parameter is never used, `ids` parameter is never used.
Consider using meaningful parameter names, like `index` -> `uuidIndex`
If your function doesn't need access to other docs, consider only passing it relevant parameters fe:
```
function processCsv(columns, row, doc, uuidIndex, toIncludeIndex)
```

File: src/fn/edit-contacts.js [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+ var row = [];
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: As far as I can tell, row is just a filtered version of `rowItem`.
Since you're not worried about mutating it (since you're doing a `splice` lower), you can use the param itself without the additional assignment.

File: src/fn/edit-contacts.js [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+ var cols = columnNames;
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: As far as I can tell, you can use the parameter itself without reassigning it to another variable.

File: src/fn/edit-contacts.js [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+ const contactId = rowItem[index];
+ const doc = contactDocs[contactId];
+
+ if(toIncludeIndex.length > 0){
+  toIncludeIndex.forEach(function(index){
+    row.push(rowItem[index]);
+  });
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: This can be rewritten as:
```
row = toIncludeIndex.map(idx => rowItem[idx]);
```

File: src/fn/edit-contacts.js [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+  toDocs.setCol(doc, col, val);
+    if(excluded) model.exclusions.push({
+      doc: doc,
+      propertyName: col,
+    });
+  }
+  return doc;
+}
+
+const fetch = {
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: Were you planning to add more properties to fetch? otherwise, the `fetchContactList` function can be standalone.

File: src/fn/edit-contacts.js [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+   const contactDocs = await db.allDocs({
+      keys: ids,
+      include_docs: true,
+   });
+    
+   const missingContactErrors = contactDocs.rows.filter(row => !row.doc).map(row => `Contact with id '${row.key}' could not be found.`);
+    if (missingContactErrors.length > 0) {
+      throw Error(missingContactErrors);
+    }
+
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: If the intent is to just process contacts, there should be a check somewhere that the requested files are indeed contacts, by checking their `type` field.
As it is now, the script can edit **any doc**, in direct opposition to its description.

**Comment by rosteve** [comment]
Hey @dianabarsan , We've now fixed all the issues. Also thanks, your review has really helped us know how to simplify code even much better and how to write better js code : ).  Would you mind to have a look at the code again?
We're now writing some tests for this.

File: src/fn/edit-contacts.js (hunk 1 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: I agree. We now read one file at a time instead of `recurseFiles` and defaults to `contact.csv` if not specified.
We've also allowed users to specify a list of comma-separated files they would wish to be processed the same way through `--files=contact.1.csv, contact.2.csv` and process them at once.

File: src/fn/edit-contacts.js (hunk 2 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Our bad. This has now been removed.

File: src/fn/edit-contacts.js (hunk 3 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Makes sense. This has now been removed.

File: src/fn/edit-contacts.js (hunk 6 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done

File: src/fn/edit-contacts.js (hunk 5 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done

File: src/fn/edit-contacts.js (hunk 4 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done

File: src/fn/edit-contacts.js (hunk 7 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done

File: src/fn/edit-contacts.js (hunk 8 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done

File: src/fn/edit-contacts.js (hunk 9 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Thanks, doing it like this makes the code much more readable 💯

File: src/fn/edit-contacts.js (hunk 10 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done

File: src/fn/edit-contacts.js (hunk 11 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Nice catch, this would've been a serious bug

File: src/fn/edit-contacts.js (hunk 12 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done

File: src/fn/edit-contacts.js (hunk 13 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: yeah, this was an unnecessary assignment. The unnecessary assignment has now been removed.

File: src/fn/edit-contacts.js (hunk 14 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done.

File: src/fn/edit-contacts.js (hunk 15 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done.

File: src/fn/edit-contacts.js (hunk 16 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Yeah, we were planning to also process user-settings documents, but we've opted to use reuse the code. We've also made the `fetchContactList ` to be standalone since we opted to reuse it.

File: README.md [hunk 18]
```diff
@@ -124,6 +124,32 @@ The list of available actions can be seen via `medic-conf --help`.
 * upload JSON files as docs on instance
 * compress PNGs and SVGs in the current directory and its subdirectories
 
+## Editing contacts across the hierarchy. 
+To edit existing couchdb documents, create a CSV file that contains the id's of the document you wish to update, and the columns of the document attribute(s) you wish to add/edit. By default, values are parsed as strings. To parse a CSV column as a JSON type, refer to the [Property Types](#property-types) section to see how you can parse the values to different types. Also refer to the [Excluded Columns](#excluded-columns) section to see how to exclude column(s) from being added to the docs.
+
+Parameter | Description | Required 
+-- | -- | --
+column(s) | Comma delimited list of columns you wish to add/edit. If this is not specified all columns will be added. | No
+docDirectoryPath | This action outputs files to local disk at this destination | No. Default `json-docs`
```
Comment (dianabarsan) [inline README.md]: You should also document the new argument `file`/`files`

File: README.md [hunk 19]
```diff
@@ -124,6 +124,32 @@ The list of available actions can be seen via `medic-conf --help`.
 * upload JSON files as docs on instance
 * compress PNGs and SVGs in the current directory and its subdirectories
 
+## Editing contacts across the hierarchy.
```
[the remark this answers is not in the document]
Comment (IreneAchola) [reply README.md]: We're looking to update contacts and users only.

File: src/fn/edit-contacts.js [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+  } else {
+    if (!columnsAreValid(cols,inputColumns)) {
+      throw Error('The column name(s) specified do not exist.');
+    }
+
+    inputColumns = cols.filter(e => inputColumns.includes(e.split(':')[0]));
+    toIncludeIndex = inputColumns.map(column => cols.indexOf(column));
+
+    if (inputColumns.includes('documentID')) {
```
[the remark this answers is not in the document]
Comment (rosteve) [reply src/fn/edit-contacts.js]: `toIncludeIndex` contains indexes of the columns and not the column titles hence the reason why need to check and get its index from `inputColumns` now renamed to `toIncludeColumns` by 


`toIncludeColumns.splice(toIncludeColumns.indexOf(DOCUMENT_ID),1);`

File: src/fn/edit-contacts.js (hunk 17 above)
Comment (IreneAchola) [reply src/fn/edit-contacts.js]: Done

File: README.md (hunk 18 above)
Comment (rosteve) [reply README.md]: Done.

File: src/fn/edit-contacts.js [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+
+  if(toIncludeIndex.length > 0){
+    row = toIncludeIndex.map(uuidIndex => row[uuidIndex]);
+  } else {
+    row.splice(uuidIndex,1);
+  }
+
+  for(let i=0; i<cols.length; ++i) {
+    const { col, val, excluded } = toDocs.parseColumn(cols[i], row[i]);
+    const colParts = col.split('.');
```
[the remark this answers is not in the document]
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done.

File: src/fn/edit-contacts.js [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+    
+    if(EDIT_RESERVED_COL_NAMES.includes(colParts[0]))
+      throw new Error(`Cannot set property defined by column '${col}' - this property name is protected.`);
+
+    toDocs.setCol(doc, col, val);
+    if(excluded) model.exclusions.push({
```
[the remark this answers is not in the document]
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done.

File: src/fn/edit-contacts.js [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+    row = toIncludeIndex.map(uuidIndex => row[uuidIndex]);
+  } else {
+    row.splice(uuidIndex,1);
+  }
+
+  for(let i=0; i<cols.length; ++i) {
+    const { col, val, excluded } = toDocs.parseColumn(cols[i], row[i]);
+    const colParts = col.split('.');
+    
+    if(EDIT_RESERVED_COL_NAMES.includes(colParts[0]))
```
[the remark this answers is not in the document]
Comment (rosteve) [reply src/fn/edit-contacts.js]: Done.

File: src/fn/edit-contacts.js [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  const idPrefix =  docType === 'contact' ? '' : 'org.couchdb.user:';
+  return rows.map((item) => idPrefix + item[index]);
+}
+
+function processDocs(docType, csv, contactDocs, args) {
+  const { rows, cols } = fs.readCsv(csv);
+  const uuidIndex = cols.indexOf('documentID');
+  let inputColumns = args.colNames;
```
[the remark this answers is not in the document]
Comment (rosteve) [reply src/fn/edit-contacts.js]: Yep, this has now been taken care

File: src/fn/edit-contacts.js [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+  if (!inputColumns.length) {
+    warn(' No columns specified, the script will add all the columns in the CSV!');
+    inputColumns = cols;
+
+  } else {
+    if (!columnsAreValid(cols,inputColumns)) {
+      throw Error('The column name(s) specified do not exist.');
+    }
+
+    inputColumns = cols.filter(e => inputColumns.includes(e.split(':')[0]));
```
[the remark this answers is not in the document]
Comment (rosteve) [reply src/fn/edit-contacts.js]: Sure, we've now updated the name

**Comment by garethbowen** [comment]
@dianabarsan Do you have time to re-review this one?

File: src/fn/edit-contacts.js [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+    warn(' No columns specified, the script will add all the columns in the CSV!');
+    toIncludeColumns = cols;
+    toIncludeIndex = [];
+
+  } else {
+    if (!columnsAreValid(cols,colNames)) {
+      throw Error('The column name(s) specified do not exist.');
+    }
+
+    toIncludeColumns = cols.filter(column => toIncludeColumns.includes(column.split(':')[0]));
```
Comment (dianabarsan) [inline src/fn/edit-contacts.js]: This line throws an error because you're calling `includes` for an `undefined` value.

File: test/fn/edit-contacts.spec.js [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+
+  before(async () => {
+    await uploadDocuments(docs);
+    sinon.stub(environment, 'pathToProject').get(() => editContactsPath);
+    const pouchDb = sinon.stub();
+    pouchDb.returns(pouch);
+    editContactsModule.__set__('pouch', pouchDb);
+  });
+
+  after(async () => pouch.destroy());
```
Comment (dianabarsan) [inline test/fn/edit-contacts.spec.js]: `before` and `after` run just once, before and after the whole suite. 
I think you should do the tear-up and tear-done for every test, using `beforeEach` and `afterEach`.
Also, make sure to clear the contents of `saveDocsDir` and empty your database!

File: test/fn/edit-contacts.spec.js [hunk 28]
```diff
@@ -0,0 +1,154 @@
...
+    }
+  }); 
+
+  it(`should fail when protected column names are provided`, async function(){
+
+    const docs = await fetchDocuments();
+    const processCsv = editContactsModule.__get__('processCsv');
+    
+    try {
+      processCsv('contact',['parent'],['0ebca32d-c1b7-5522-94a3-97dd8b3df146','parent_id'],0, [],docs);
```
Comment (dianabarsan) [inline test/fn/edit-contacts.spec.js]: I think this test - and all the following tests - should call `editContactsModule.execute` instead of a rewired function call. 
It's fine to have fine testing of subcomponents, but you should prioritize testing the module as a whole. 
This test doesn't answer the question of: what happens when I pass a protected column name to the module? Does it fail? I don't know. `processCsv` fails, but is it caught? does execution still happen? are there any files created?

File: test/fn/edit-contacts.spec.js [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+      const docs = await fetchDocuments();
+      
+      const parsedDocuments = docs.rows.reduce((agg, curr) => Object.assign(agg, { [curr.doc._id]: curr.doc }), {});
+
+      // Calling processDocs in edit-contacts to test if it edits documents well.
+      const result = processDocuments('contact', csv, parsedDocuments, {colNames: ''});
+
+      // Saving edited docs to disk 
+      // then comparing between the expected JSON files and generated JSON files that have been processed using edit-contacts
+      Object.values(result).map(saveJsonDoc);
```
Comment (dianabarsan) [inline test/fn/edit-contacts.spec.js]: Please refrain from re-writing module code within the test. 
You could rewrite this and run the whole suite (including saving the docs to disk) and compare the actual results.

File: src/fn/edit-contacts.js (hunk 26 above)
Comment (rosteve) [reply src/fn/edit-contacts.js]: Nice catch, This should now be fixed.

File: test/fn/edit-contacts.spec.js (hunk 27 above)
Comment (rosteve) [reply test/fn/edit-contacts.spec.js]: Done.

File: test/fn/edit-contacts.spec.js (hunk 28 above)
Comment (rosteve) [reply test/fn/edit-contacts.spec.js]: This totally makes sense, we've now implemented the tests to call `editContactsModule.execute` instead of a rewired function call so that we can test the module as a whole.

File: test/fn/edit-contacts.spec.js (hunk 29 above)
Comment (rosteve) [reply test/fn/edit-contacts.spec.js]: Done.

**Comment by rosteve** [comment]
@dianabarsan We've now done all the changes. Would you mind to re-review again?

Also, Travis is failing with some `syntax-error` issue but the local tests are working perfectly. Would you mind to help us point out what might be causing Travis to fail?

File: test/fn/edit-contacts.spec.js [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+  it(`should only add specified column names to the json docs`, async function(){
+
+    const parseResult = {
+      colNames: ['is_in_emnch'],
+      csvFiles: ['contact.csv'],
+      docDirectoryPath: path.resolve(environment.pathToProject, 'json_docs'),
+    };
+    const extraArgs = sinon.stub();
+    extraArgs.returns(parseResult);
+    editContactsModule.__set__('parseExtraArgs', extraArgs);
```
Comment (dianabarsan) [inline test/fn/edit-contacts.spec.js]: Please refrain from __set__-ing any "moving" parts of your module, especially if you're not testing them somewhere else (which you're not, in this case).
Replace every time you `editContactsModule.__set__('parseExtraArgs'` with something like:
```
sinon.stub(environment, 'extraArgs').get(() => ['--columns=is_in_emnch', '--files=contact.csv']);
```

Applies to all other tests as well.

File: test/fn/edit-contacts.spec.js [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+
+  it(`should fail when document type is not a contact`, async function(){
+
+    await pouch.put({
+      _id: 'documentID',
+      type: 'data_record'
+    });
+    const getIDs = sinon.stub();
+    getIDs.returns(['documentID']);
+    editContactsModule.__set__('getIDs', getIDs);
```
Comment (dianabarsan) [inline test/fn/edit-contacts.spec.js]: Please don't `__set__` `getIDs` here. Rather, setup your database and test csv so your module tries to edit an inexistent doc. Same goes for the other test.

File: test/fn/edit-contacts.spec.js [hunk 32]
```diff
@@ -0,0 +1,197 @@
+const { expect, assert } = require('chai');
+const rewire = require('rewire');
+const PouchDB = require('pouchdb-core');
+PouchDB.plugin(require('pouchdb-adapter-memory'));
+const fs = require('../../src/lib/sync-fs');
+const environment = require('../../src/lib/environment');
+const sinon = require('sinon');
+// const info = require('../../src/lib/log');
```
Comment (dianabarsan) [inline test/fn/edit-contacts.spec.js]: Please remove this comment.

File: test/fn/edit-contacts.spec.js [hunk 33]
```diff
@@ -0,0 +1,197 @@
...
+  .map(file => {
+    const doc = fs.readJson(file);
+    return doc;
+  });
+
+const uploadDocuments = (docs) => {
+  return pouch.bulkDocs(docs);
+};
+
+function compareDocuments(expectedDocsDir){
```
Comment (dianabarsan) [inline test/fn/edit-contacts.spec.js]: consider changing this to use `const <function_name> = ....`, for consistency :)
